debian/messaging-app-apparmor.manifest

```diff
@@ -27,8 +27,10 @@
         "/usr/share/applications/",
         "/custom/xdg/data/dconf/",
         "/usr/share/*/assets/",
+        "@{PROC}/[0-9]*/task/[0-9]*/comm",
         "@{HOME}/.local/share/evolution/addressbook/*/photos/",
         "@{HOME}/.cache/messaging-app/HubIncoming/**",
+        "@{HOME}/.cache/qtshadercache-arm64-little_endian-lp64/**",
```

Comment:
Note the documentation for https://doc.qt.io/qt-5/qopenglshaderprogram.html#caching-program-binaries states the following:

> this will transparently cache program binaries under [QStandardPaths::GenericCacheLocation](https://doc.qt.io/qt-5/qstandardpaths.html#StandardLocation-enum) or [QStandardPaths::CacheLocation](https://doc.qt.io/qt-5/qstandardpaths.html#StandardLocation-enum).

`QStandardPaths::GenericCacheLocation` is "a directory location where user-specific non-essential (cached) data, shared across applications, should be written." `QStandardPaths::CacheLocation` is "a directory location where user-specific non-essential (cached) data should be written. This is an application-specific directory."

Given there are not two apparmor denials in sequence, I would assume that the cache function is falling back to the application-specific cache directory. However, assumptions are dangerous. Since Qt is open source, we can confirm our assumptions by looking at the code.

If we start at [`QOpenGLShaderProgram::addCacheableShaderFromSourceCode(...)`](https://code.woboq.org/qt5/qtbase/src/gui/opengl/qopenglshaderprogram.cpp.html#_ZN20QOpenGLShaderProgram32addCacheableShaderFromSourceCodeE6QFlagsIN13QOpenGLShader13ShaderTypeBitEERK10QByteArray), we can see that the compilation step is `d->binaryProgram.shaders.append(QOpenGLProgramBinaryCache::ShaderDesc(type, source));`. `d->binaryProgram.shaders` seems interesting, especially since we're adding one. Let's follow `shaders` back and find that it is a [`QOpenGLProgramBinaryCache`'s `QVector<ShaderDesc>`, which means it can be appended to](https://code.woboq.org/qt5/qtbase/src/gui/opengl/qopenglprogrambinarycache_p.h.html#QOpenGLProgramBinaryCache::ProgramDesc::shaders). But `QOpenGLProgramBinaryCache` also has an interesting `save(...)` method. [Let's follow that back](https://code.woboq.org/qt5/qtbase/src/gui/opengl/qopenglprogrambinarycache.cpp.html#_ZN25QOpenGLProgramBinaryCache4saveERK10QByteArrayj). This method seems to do some workarounds for shader programs being loaded incorrectly on cache hits, then saves the files:

https://github.com/qt/qtbase/blob/823ed71e220ebde07970dd61c04bb47b01dd06c4/src/gui/opengl/qopenglprogrambinarycache.cpp#L416-L423

[`cacheFileName`](https://github.com/qt/qtbase/blob/823ed71e220ebde07970dd61c04bb47b01dd06c4/src/gui/opengl/qopenglprogrambinarycache.cpp#L119-L122) is `m_cacheDir + QString::fromUtf8(cacheKey);`. `m_cacheDir` is set by [`QOpenGLProgramBinaryCache::QOpenGLProgramBinaryCache()`](https://github.com/qt/qtbase/blob/823ed71e220ebde07970dd61c04bb47b01dd06c4/src/gui/opengl/qopenglprogrambinarycache.cpp#L103-L117). Here we see an interesting `qCDebug`, telling us that we should be able to get the debug output by enabling a logging category by the name of `DBG_SHADER_CACHE`. Strangely, we find it on [`qopenglshaderprogram.cpp:182`](https://github.com/qt/qtbase/blob/v5.12.9/src/gui/opengl/qopenglshaderprogram.cpp#L182), which shows us that the logging category is `"qt.opengl.diskcache"`. Now we can enable logging for this category:

```
phablet@ubuntu-phablet:~$ initctl set-env QT_LOGGING_RULES='qt.opengl.*=true'
```

Then run messaging-app again. We see the following output in the logs:

```
qt.opengl.diskcache: Cache location '/home/phablet/.cache/qtshadercache-arm64-little_endian-lp64/' writable = 1
qt.opengl.diskcache: program with 2 shaders, cache key 7d5f9fc417d36376dbfbebf864483aca3a25ac1a
qt.opengl.diskcache: Program binary not in cache, compiling
qt.opengl.diskcache: Program binary is 12356 bytes, err = 0x0, total 12480
qt.opengl.diskcache: Failed to create /home/phablet/.cache/qtshadercache-arm64-little_endian-lp64/7d5f9fc417d36376dbfbebf864483aca3a25ac1a in shader cache
```

So, the problem could be that Qt is trying to use the global shader cache directory when it should be using the process-specific directory. Getting Qt to use the process-specific directory would be my goal now if I were you, to prevent malicious applications from doing something nefarious with the global shader cache that would cause bad behavior in other apps. However unlikely that may be, I think it's a possible threat and one we should shield ourselves from.

How you get Qt to use the process-specific cache is an exercise left to you.

Reply:
Well as of the source code that we see that was not foreseen to be configurable. The question is do we want to patch that over, without knowing what was the rationale? Or is this smth that should be opened as an issue first with Qt, to see what they think, I would go for this first.

Reply:
I agree, a Qt bug would be helpful. Qt is supposed to auto-detect whether the default location is writable, and we appear to have found a case where the default location is **not** writable but Qt doesn't detect it.

Reply:
Filed as https://bugreports.qt.io/browse/QTBUG-96789. Review at https://codereview.qt-project.org/c/qt/qtbase/+/373658. Backport at https://gitlab.com/ubports/core/packaging/qtbase/-/merge_requests/2.